File: spark/src/main/scala/geotrellis/spark/io/index/zcurve/ZSpatialKeyIndex.scala
package geotrellis.spark.io.index.zcurve

import geotrellis.spark._
import geotrellis.spark.io.index._

class ZSpatialKeyIndex() extends KeyIndex[SpatialKey] {
  private def toZ(key: SpatialKey): Z2 = Z2(key.col, key.row)

  def toIndex(key: SpatialKey): Long = toZ(key).z

  def indexRanges(keyRange: (SpatialKey, SpatialKey)): Seq[(Long, Long)] =
    Z2.zranges(toZ(keyRange._1), toZ(keyRange._2))
}

File: geotrellis-python/geotrellis/spark/io/index/zcurve/ZSpatialKeyIndex.py
from geotrellis.spark.io.index.KeyIndex import KeyIndex
from geotrellis.spark.SpatialKey import SpatialKey
from geotrellis.spark.io.json.KeyIndexFormats import ZSpatialKeyIndexFormat
from Z2 import Z2

class ZSpatialKeyIndex(KeyIndex[SpatialKey]):
    implicits = {'format': lambda: ZSpatialKeyIndexFormat()}
    def __init__(self, keyBounds):
        self._keyBounds = keyBounds
    @property
    def keyBounds(self):
        return self._keyBounds
    def _toz(self, key):
        return Z2(key.col, key.row)
    def toIndex(self, key):
        return self._toz(key).z
    def indexRanges(self, keyRange):
        return Z2.zranges(self._toz(keyRange[0]), self._toz(keyRange[1]))

